src/Analyzers/MSTest.Analyzers/DataRowShouldBeValidAnalyzer.cs

```diff
@@ -0,0 +1,214 @@
+        {
+            return;
+        }
+
+        // Count mismatch since there's zero method parameters but there's at least one DataRow
+        // constructor argument.
+        if (methodSymbol.Parameters.Length == 0)
+        {
+            context.ReportDiagnostic(syntax.CreateDiagnostic(
+                ArgumentCountMismatchRule,
+                attribute.ConstructorArguments.Length,
+                methodSymbol.Parameters.Length));
+            return;
+        }
+
+        // Possible count mismatch depending on whether last method parameter is an array or not.
+        IParameterSymbol lastMethodParameter = methodSymbol.Parameters.Last();
+        bool lastMethodParameterIsArray = lastMethodParameter.Type.Kind == SymbolKind.ArrayType;
+        if (attribute.ConstructorArguments.Length == 0)
+        {
+            if (!lastMethodParameterIsArray)
+            {
+                context.ReportDiagnostic(syntax.CreateDiagnostic(
+                    ArgumentCountMismatchRule,
+                    attribute.ConstructorArguments.Length,
+                    methodSymbol.Parameters.Length));
+            }
+
+            return;
+        }
+
+        // DataRow constructors have either zero or one argument(s). If we get here, we are
+        // on the one argument case. Check if we match either of the array argument constructors
+        // and expand the array argument if we do.
+        ImmutableArray<TypedConstant> constructorArguments = attribute.ConstructorArguments;
+        if (constructorArguments[0].Kind is TypedConstantKind.Array && !constructorArguments[0].IsNull)
+        {
+            constructorArguments = constructorArguments[0].Values;
+        }
+
+        if (IsArgumentCountMismatch(constructorArguments.Length, methodSymbol.Parameters.Length, lastMethodParameterIsArray))
+        {
+            context.ReportDiagnostic(syntax.CreateDiagnostic(
+                ArgumentCountMismatchRule,
+                constructorArguments.Length,
+                methodSymbol.Parameters.Length));
+            return;
+        }
+
+        // Check constructor argument types match method parameter types.
+        List<(int ConstructorArgumentIndex, int MethodParameterIndex)> typeMismatchIndices = new();
+        for (int i = 0, j = 0; i < constructorArguments.Length; ++i, ++j)
+        {
+            if (constructorArguments[i].IsNull)
+            {
+                continue;
+            }
+
+            ITypeSymbol? argumentType = constructorArguments[i].Type;
+            ITypeSymbol paramType = (lastMethodParameterIsArray && j >= methodSymbol.Parameters.Length - 1)
+                ? ((IArrayTypeSymbol)lastMethodParameter.Type).ElementType
+                : methodSymbol.Parameters[j].Type;
+
+            if (argumentType is not null && !argumentType.IsAssignableTo(paramType, context.Compilation))
+            {
+                typeMismatchIndices.Add((i, j));
```

Comment:
These 2 indices are always the same in your alg so there isn't much point in having both, or am I missing something?

Reply:
They are not necessarily the same, the lists could still have different lengths if the last method parameter is an array and there are more ctor args (e.g.: 4 ctor args, 2 method params and last is an actual params array). However, in that case I should be able to use the actual index of the last param.